keps/sig-auth/20191101-add-request-id-to-k8s-logs.md

```diff
@@ -0,0 +1,308 @@
+
+It is difficult that support team in k8s Service Provider resolve end user's problem quickly in the above. Therefore, we'd like to add a new identifier which is unique to each user operation. This feature is useful for the following use cases:
+
+#### Case 1
+
+In case of insecure or unauthorized operation happens, it is necessary to
+identify what effect that operation caused. This proposed feature helps identify
+what happened at each component or server by each insecure / unauthorized API
+request. We can collect these logs as an evidence.
+
+#### Case 2
+
+If the container is terminated by OOM killer, there is a case to break down the
+issue into parts(Pod or Kubernetes) from the messages related OOM killer on host logs
+and the API processing just before OOM killer. If the cause is that some unknown
+pod creations, it is helpful to detect the root API request and who called this
+request.
+
+### Goals
+
+ - Adding a Request-ID into each K8s component log.
+ - The Request-ID is unique to an operation.
+ - Control enabled/disabled Request-ID feature(Request-ID feature is disabled on default to avoid an impact for existing user).
+
+### Non-Goals
+
+ - To centrally manage the logs of each Kubernetes component with Request-ID (This can
+be realized with existing OSS such as Kibana, so no need to implement into Kubernetes
+components).
+ - We don't associate Request-ID to all of operations(Our target is important operations such as `kubectl create/delete/etc.`).
+
+## Proposal
+
+### Basic policy of Request-ID
+
+ - Minimize the impact to existing users who are retrieving logs and analyzing with existing log format.
+   - So we disabled Request-ID feature on default.
+ - Collaborate with related KEPs to avoid unnecessary conflict to them regarding implementation and feature.
+   - Use existing KEP's feature as much as possible.
+   - Therefore, we will merge Request-ID feature after related KEP features are merged.
+
+### Design Overview
+
+Request-ID feature consists the three features.
+ - Propagate Request-ID to related objects.
+ - Export Request-ID to each Kubernetes component log.
+ - Control enabled/disabled exporting Request-ID.
+
+#### Design of Propagate Request-ID
+
+There is an idea to use `distributed context` of the existing KEP([Distributed Tracing](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/0034-distributed-tracing-kep.md)). We use `Distributed Tracing` feature for propagation.
+
+ - We just use Tracing codes provided by `Distributed Tracing` feature.
+ - We use Trace-ID as Request-ID (This does not interfere Tracing codes).
+
+The target functions that we add Request-id codes are the following tables. The target functions include the klog call which called via important `kubectl` operations. These operations are listed in [Migration / Graduation Criteria](#migration--graduation-criteria) section. As a result, we can propagate Request-ID regarding important `kubectl` operations.
```

Comment:
What was the criteria for picking those and not others?

Reply:
What marks an `important kubectl operation`?

Reply:
I added criteria and explanation in the KEP.